ggml-metal.metal

```diff
@@ -400,20 +436,15 @@ kernel void kernel_mul_mat_q4_0_f32(
             y_curr[i] = *((device float4  *)(y + N_SIMDWIDTH * (tiisg + column * QK4_0) + 4 * i));
             sumy += y_curr[i][0] + y_curr[i][1] + y_curr[i][2] + y_curr[i][3];
         }
-        sumy *= (-8.f);
+        // we don't right shift packed 4-bit weights, so we have to devide y by 16/256/4096 to conpensate this.
+        // this design is q4_0 and q4_1 centered, but I think most of the people use these two quantizations.
```

Comment:
Many people do use quantizations other than `Q4_0` and `Q4_1`.

Reply:
Sorry for the misunderstanding, I mean among `Q4_0`, `Q4_1`, `Q5_0`, `Q5_1` and `Q8_0` people mostly use `Q4_0` and `Q4_1`. I think `Q5` can also benefits a little from this design, and only for `Q8` we need to do 4 more multiplications for each block.
I know that quite a lot people use k-quants, and for myself I also use k-quants model mostly for it provides better perplexity. This template mainly serves for non-k quants, because k-quants use much larger block size, and may need a different load strategy. Once I finish optimize this template, I will try to apply the same design rules to k-quants and see if we can merge all quantization methods together or we have to prepare a separate template for k-quants.